src/devices/src/virtio/file_traits.rs

```diff
@@ -411,3 +411,49 @@ macro_rules! volatile_impl {
 }
 
 volatile_impl!(File);
+
+impl FileReadWriteAtVolatile for imago::SyncFormatAccess<imago::file::File> {
+    fn read_at_volatile(&self, slice: VolatileSlice, offset: u64) -> Result<usize> {
+        let slices = &[&slice];
+        let iovec = imago::io_buffers::IoVectorMut::from_volatile_slice(slices);
+        self.readv(iovec.0, offset)?;
+        Ok(slice.len())
+    }
+
+    fn read_vectored_at_volatile(&self, bufs: &[VolatileSlice], offset: u64) -> Result<usize> {
+        if bufs.is_empty() {
+            return Ok(0);
+        }
+
+        let mut offset = offset as usize;
+        let mut read = 0;
+        for buf in bufs {
+            let bytes = self.read_at_volatile(*buf, offset as u64)?;
+            offset += bytes;
+            read += bytes;
+        }
+        Ok(read)
```

Comment:
I think this should just be
```rust
let (iovec, _guard) = imago::io_buffers::IoVectorMut::from_volatile_slice(bufs);
let full_length = iovec.len();
self.readv(iovec, offset)?;
Ok(full_length)
```

But I see that libkrun has `&[VolatileSlice]` here, whereas virtiofsd had `&[&VolatileSlice]`, so I implemented the latter (because that’s what I used for testing)…

I’ll add an interface specifically for `&[VS]`. I’m trying to allow the existing `from_volatile_slice()` to accept it (I hope I can get something done with `IntoIter<Item = &'a VS>`), but if it turns out not to be possible, I’ll add a `from_owned_volatile_slice()` variant.

Reply:
I was able to make the interface accept either (basically `IntoIterator<Item: AsRef<VolatileSlice>, IntoIter: ExactSizeIterator>`, to be specific (with a custom `AsRef`)), so the code I put above should now work.

Reply:
Also should be fixed.